### ISSUE TITLE
DNS resolution iptables rules that are set inside the container are invalid

### ISSUE THREAD

**Issue by chrisUsick** [moby/moby#40428]
**Description**

As of docker engine 19.03.5 the firewall rules injected into the container at startup time are not valid iptables rules for `iptables-restore`. When the container starts up they seem to work but if you remove them, then try adding them back (for example, because firewalld starts) then the rules can not be restored and thus, DNS resolution because broken on that container. 

**Steps to reproduce the issue:**

1. create a docker bridge network: `docker network create -d bridge dokken`
1. Start a `centos 7` docker container: 
  ```
  export ID=$(docker run -d --network dokken -v /sys/fs/cgroup:/sys/fs/cgroup --privileged --rm -it dokken/centos-7 /usr/sbin/init)
  ```
1. Confirm dns is resolving: `docker exec -it $ID /bin/bash -c 'curl -Lv http://www.google.com 1>/dev/null'`
1. save iptables to text file: `docker exec -it $ID /bin/bash -c 'iptables-save -c > /tmp/docker-iptables.txt'`
1. Drop firewall rules
  - This will simulate having firewalld start at startup time, which also clears away docker iptables rules
  - `docker exec -it $ID /bin/bash -c 'iptables -F -t nat'`
1. restore iptables rules from text file: `docker exec -it $ID /bin/bash -c 'iptables-restore -n -c /tmp/docker-iptables.txt'`

**Describe the results you received:**

```
iptables-restore v1.4.21: Port `0' not valid

Error occurred at line: 18
Try `iptables-restore -h' or 'iptables-restore --help' for more information.
```

**Describe the results you expected:**

`iptables-restore` to succeed. 

**Additional information you deem important (e.g. issue happens only occasionally):**

Before I upgrade docker for desktop (see below for my complete environment) docker created firewall rules like the following:

```
# Generated by iptables-save v1.4.21 on Fri Dec 20 04:20:50 2019
*filter
:INPUT ACCEPT [69:5664]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [69:5536]
COMMIT
# Completed on Fri Dec 20 04:20:50 2019
# Generated by iptables-save v1.4.21 on Fri Dec 20 04:20:50 2019
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [14:1002]
:POSTROUTING ACCEPT [26:1796]
:DOCKER_OUTPUT - [0:0]
:DOCKER_POSTROUTING - [0:0]
[12:794] -A OUTPUT -d 127.0.0.11/32 -j DOCKER_OUTPUT
[12:794] -A POSTROUTING -d 127.0.0.11/32 -j DOCKER_POSTROUTING
[0:0] -A DOCKER_OUTPUT -d 127.0.0.11/32 -p tcp -m tcp --dport 53 -j DNAT --to-destination 127.0.0.11:38159
[12:794] -A DOCKER_OUTPUT -d 127.0.0.11/32 -p udp -m udp --dport 53 -j DNAT --to-destination 127.0.0.11:57554
[0:0] -A DOCKER_POSTROUTING -s 127.0.0.11/32 -p tcp -m tcp --sport 38159 -j SNAT --to-source :53
[0:0] -A DOCKER_POSTROUTING -s 127.0.0.11/32 -p udp -m udp --sport 57554 -j SNAT --to-source :53
COMMIT
# Completed on Fri Dec 20 04:20:50 2019
```

After the upgrade the firewall rules created by docker look like the following: 

```
# Generated by iptables-save v1.4.21 on Wed Jan 29 16:38:14 2020
*filter
:INPUT ACCEPT [20:14206]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [21:1102]
COMMIT
# Completed on Wed Jan 29 16:38:14 2020
# Generated by iptables-save v1.4.21 on Wed Jan 29 16:38:14 2020
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [3:180]
:POSTROUTING ACCEPT [4:240]
:DOCKER_OUTPUT - [0:0]
:DOCKER_POSTROUTING - [0:0]
[1:60] -A OUTPUT -d 127.0.0.11/32 -j DOCKER_OUTPUT
[1:60] -A POSTROUTING -d 127.0.0.11/32 -j DOCKER_POSTROUTING
[0:0] -A DOCKER_OUTPUT -d 127.0.0.11/32 -p tcp -m tcp --dport 53 -j DNAT --to-destination 0.0.0.0:0 --random --persistent --to-destination 0.0.0.0:0 --random --persistent --to-destination :0
[1:60] -A DOCKER_OUTPUT -d 127.0.0.11/32 -p udp -m udp --dport 53 -j DNAT --to-destination 0.0.0.0:0 --random --persistent --to-destination 0.0.0.0:0 --random --persistent --to-destination 0.0.0.0:0
[0:0] -A DOCKER_POSTROUTING -s 127.0.0.11/32 -p tcp -m tcp --sport 37395 -j SNAT --to-source :53
[0:0] -A DOCKER_POSTROUTING -s 127.0.0.11/32 -p udp -m udp --sport 58114 -j SNAT --to-source :53
COMMIT
# Completed on Wed Jan 29 16:38:14 2020
```

_Note:_ support for multiple `--to-destination` options was dropped in kernels `>=2.6.11-rc1` ([iptables manual](https://linux.die.net/man/8/iptables))
Docker for desktop uses a linux kernel of `4.19.76` ([release notes](https://docs.docker.com/docker-for-mac/release-notes/))
It also appears invalid to use `0` as a port in a `--to-destination` option. 

Unfortunately I don't know which docker engine version I had before I upgraded docker for desktop but it seems that the upgrade changed things. I upgraded docker desktop while I was in the middle of working on firewall rules, and they immediately changed from the old to the new format.
I'm further confused by the fact that I can't identify where the code changed.  It appears that https://github.com/docker/libnetwork/blob/master/iptables/iptables.go is responsible for creating these rules but it hasn't changed since April, 2019.

I'm not sure why docker can insert the firewall rules initially but not restore them. 

My short term workaround is going to be to extract the `-sport` value out of the `DOCKER_POSTROUTING` rules, and craft valid iptables rules before restoring them.
In the long term, if I'm not doing something wrong here and the new iptables rules are actually invalid, it could become an issue down the road. 
**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea
 Built:             Wed Nov 13 07:22:34 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea
  Built:            Wed Nov 13 07:29:19 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 7
  Running: 2
  Paused: 0
  Stopped: 5
 Images: 430
 Server Version: 19.03.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 1.943GiB
 Name: docker-desktop
 ID: MHIR:VZ7H:3C6P:JT2H:KN35:XAFD:XMJA:B7YD:PB2X:S7YR:XFHF:S2HV
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 56
  Goroutines: 71
  System Time: 2020-01-29T16:43:58.8321308Z
  EventsListeners: 3
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I am running docker desktop for Mac OS version 2.2.0.0.

**Comment by ncopa** [comment]
I can reproduce this with `alpine:3.2`, which also uses iptables-1.4.2, and with `alpine:3.4` with iptables-1.6.0. However, with iptables-1.6.1 the `iptables-save` generates different output:

```
 # iptables-save
# Generated by iptables-save v1.6.1 on Mon Feb  3 18:48:00 2020
*filter
:INPUT ACCEPT [1219:1738266]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [346:14872]
COMMIT
# Completed on Mon Feb  3 18:48:00 2020
# Generated by iptables-save v1.6.1 on Mon Feb  3 18:48:00 2020
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [3:196]
:POSTROUTING ACCEPT [4:264]
:DOCKER_OUTPUT - [0:0]
:DOCKER_POSTROUTING - [0:0]
-A OUTPUT -d 127.0.0.11/32 -j DOCKER_OUTPUT
-A POSTROUTING -d 127.0.0.11/32 -j DOCKER_POSTROUTING
-A DOCKER_OUTPUT -d 127.0.0.11/32 -p tcp -m tcp --dport 53 -j DNAT [unsupported revision]
-A DOCKER_OUTPUT -d 127.0.0.11/32 -p udp -m udp --dport 53 -j DNAT [unsupported revision]
-A DOCKER_POSTROUTING -s 127.0.0.11/32 -p tcp -m tcp --sport 40111 -j SNAT --to-source :53
-A DOCKER_POSTROUTING -s 127.0.0.11/32 -p udp -m udp --sport 53557 -j SNAT --to-source :53
COMMIT
# Completed on Mon Feb  3 18:48:00 2020
```
Note the `[unsupported revision]`.

I think the problem is that the iptables rules are configured with newer iptables and kernel and uses features that is not supported by the old iptables-1.4.2.

I believe this is a bug in iptables-1.4.2.

**Comment by chrisUsick** [comment]
Thanks for the feedback. Interesting that the iptables rules aren’t the
same as the ones I got. Could docker be doing different rules for different
kernel versions?

**Comment by cpuguy83** [comment]
Reminder that the iptables version in the container should match iptables version on the host.
Docker doesn't do different rules based on kernel version as far as I am aware (actually maybe for pre-3.16 for vxlan issues).

The differences are likely due to the host iptables actually being converted to nftables rules (by the iptables binary) while the container is still doing older iptables rules, or vice-versa.

Hope that makes sense.

TLDR make sure your iptables binaries are talking the same protocol in the container and on the host because the distros have started making iptables write rules for nftables.

**Comment by chrisUsick** [comment]
Thank you, this makes sense. I’ll try my test on a container with the same
iptables version that docker for desktop uses.

**Comment by oxr463** [comment]
Is there a security concern here? What is the worst possible outcome if the firewall rules are corrupted?

**Comment by chrisUsick** [comment]
The concern is that docker subsequent processes that rely on dns don’t
succeed. If DNS doesn’t work there are a lot of use cases that don’t work.
